src/v/archival/probe.cc

```diff
@@ -105,23 +129,57 @@ void ntp_level_probe::setup_public_metrics(const model::ntp& ntp) {
            "This may grow due to retention or non compacted segments being "
            "replaced with their compacted equivalent."),
          labels)
-         .aggregate(aggregate_labels),
-       sm::make_gauge(
-         "segments",
-         [this] { return _segments_in_manifest; },
-         sm::description(
-           "Total number of accounted segments in the cloud for the topic"),
-         labels)
-         .aggregate(aggregate_labels),
-       sm::make_gauge(
-         "segments_pending_deletion",
-         [this] { return _segments_to_delete; },
-         sm::description("Total number of segments pending deletion from the "
-                         "cloud for the topic"),
-         labels)
          .aggregate(aggregate_labels)});
 }
 
+void ntp_level_probe::setup_dynamic_public_metrics(const model::ntp& ntp) {
+    namespace sm = ss::metrics;
+
+    auto ns_label = ssx::metrics::make_namespaced_label("namespace");
+    auto topic_label = ssx::metrics::make_namespaced_label("topic");
+    auto partition_label = ssx::metrics::make_namespaced_label("partition");
+    const std::vector<sm::label_instance> labels = {
+      ns_label(ntp.ns()),
+      topic_label(ntp.tp.topic()),
+      partition_label(ntp.tp.partition()),
+    };
+
+    auto aggregate_labels = std::vector<sm::label>{
+      sm::shard_label, partition_label};
+
+    _dynamic_public_metrics->add_group(
+      prometheus_sanitize::metrics_name("cloud_storage"),
+      {
+        sm::make_gauge(
+          "segments",
+          [this] { return _segments_in_manifest; },
+          sm::description(
+            "Total number of accounted segments in the cloud for the topic"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_gauge(
+          "segments_pending_deletion",
+          [this] { return _segments_to_delete; },
+          sm::description("Total number of segments pending deletion from the "
+                          "cloud for the topic"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_gauge(
+          "records_pending_upload",
```

Comment:
I wonder what the use case is for this?  I would have thought an SRE would almost always care about the raw quantity of data to upload, rather than how many records are contained within it

Reply:
In previous incidents we tried to use this metric to determine if all user data was uploaded before performing an operation. In the previous form, the metric wasn't helpful.

Reply:
Is the idea here that it was only not useful because it wasn't a public? Might be worth noting that it is an identical version of the `pending` metric.

Reply:
In its previous form the metric would not converge to zero in some cases due to it being uploaded from inside the upload loop.